### ISSUE TITLE
Retraining with Resnet50 backbone

### ISSUE THREAD

**Issue by vladimirmujagic** [vita-epfl/openpifpaf#329]
Hi,

Can you please provide instructions (hyperparams, example train script call) on how to reproduce pretrained resnet50 model  stored on your torch-hub.
Current guide only describes how to train shufflenet.

Regards,

**Comment by svenkreiss** [comment]
Please check the first line in the corresponding log file. It contains the exact command line arguments that were used and also the exact version of OpenPifPaf.

**Comment by vladimirmujagic** [comment]
Thank you very much for quick response, I found it.

**Comment by vladimirmujagic** [comment]
Update:

I managed to run first training experiment with default parameters.

# Environment
### System Environment
- CUDA: `10.1`
- Nvidia Driver: `418.87.01`
  
### Python3.7.5 Environment
```
torch==1.4.0
torchvision==0.5.0

boto==2.49.0
certifi==2020.12.5
chardet==4.0.0
cycler==0.10.0
Cython==0.29.21
future==0.18.2
idna==2.10
kiwisolver==1.3.1
matplotlib==3.3.3
numpy==1.19.4
opencv-python==4.4.0.46
openpifpaf==0.11.9+229.g6385c33.dirty
Pillow==8.0.1
pycocotools==2.0.2
pyparsing==2.4.7
pysparkling==0.6.0
python-dateutil==2.8.1
python-json-logger==2.0.1
requests==2.25.1
scipy==1.6.0rc2
six==1.15.0
thop==0.0.31.post2005241907
typing-extensions==3.7.4.3
urllib3==1.26.2

```
### Training Setup
- Openpifpaf commit hash: `6385c33b2c32e305d045362e662d732a9dd3c78a`
- Training with default parameters
```
{
   "message":null,
   "levelname":"INFO",
   "name":"openpifpaf.logger",
   "asctime":"2020-12-25 11:23:47,722",
   "type":"process",
   "argv":[
      "/usr/local/lib/python3.7/dist-packages/openpifpaf-0.11.9+229.g6385c33.dirty-py3.7-linux-x86_64.egg/openpifpaf/train.py",
      "--basenet",
      "resnet50",
      "--dataset",
      "cocokp"
   ],
   "args":{
      "quiet":false,
      "debug":false,
      "log_stats":false,
      "shufflenetv2_pretrained":true,
      "shufflenetv2k_input_conv2_stride":0,
      "shufflenetv2k_input_conv2_outchannels":null,
      "shufflenetv2k_stage4_dilation":1,
      "shufflenetv2k_kernel":5,
      "shufflenetv2k_instance_norm":false,
      "shufflenetv2k_group_norm":false,
      "resnet_pretrained":true,
      "resnet_pool0_stride":0,
      "resnet_input_conv_stride":2,
      "resnet_input_conv2_stride":0,
      "resnet_block5_dilation":1,
      "resnet_remove_last_block":false,
      "cf3_dropout":0.0,
      "cf3_inplace_ops":true,
      "checkpoint":null,
      "basenet":"resnet50",
      "two_scale":false,
      "multi_scale":false,
      "multi_scale_hflip":true,
      "cross_talk":0.0,
      "download_progress":true,
      "head_consolidation":"filter_and_extend",
      "lambdas":null,
      "r_smooth":0.0,
      "regression_loss":"laplace",
      "background_weight":1.0,
      "b_scale":1.0,
      "focal_gamma":1.0,
      "auto_tune_mtl":false,
      "auto_tune_mtl_variance":false,
      "task_sparsity_weight":0.0,
      "cif_side_length":4,
      "caf_min_size":3,
      "caf_fixed_size":false,
      "caf_aspect_ratio":0.0,
      "encoder_suppress_selfhidden":true,
      "momentum":0.9,
      "beta2":0.999,
      "adam_eps":1e-06,
      "nesterov":true,
      "weight_decay":0.0,
      "adam":false,
      "amsgrad":false,
      "lr":0.001,
      "lr_decay":[
         
      ],
      "lr_decay_factor":0.1,
      "lr_decay_epochs":1.0,
      "lr_warm_up_start_epoch":0,
      "lr_warm_up_epochs":1,
      "lr_warm_up_factor":0.001,
      "lr_warm_restarts":[
         
      ],
      "lr_warm_restart_duration":0.5,
      "dataset":"cocokp",
      "loader_workers":null,
      "batch_size":8,
      "dataset_weights":null,
      "cocodet_train_annotations":"data-mscoco/annotations/instances_train2017.json",
      "cocodet_val_annotations":"data-mscoco/annotations/instances_val2017.json",
      "cocodet_train_image_dir":"data-mscoco/images/train2017/",
      "cocodet_val_image_dir":"data-mscoco/images/val2017/",
      "cocodet_square_edge":513,
      "cocodet_extended_scale":false,
      "cocodet_orientation_invariant":0.0,
      "cocodet_blur":0.0,
      "cocodet_augmentation":true,
      "cocodet_rescale_images":1.0,
      "cocodet_upsample":1,
      "cocokp_train_annotations":"data-mscoco/annotations/person_keypoints_train2017.json",
      "cocokp_val_annotations":"data-mscoco/annotations/person_keypoints_val2017.json",
      "cocokp_train_image_dir":"data-mscoco/images/train2017/",
      "cocokp_val_image_dir":"data-mscoco/images/val2017/",
      "cocokp_square_edge":385,
      "cocokp_extended_scale":false,
      "cocokp_orientation_invariant":0.0,
      "cocokp_blur":0.0,
      "cocokp_augmentation":true,
      "cocokp_rescale_images":1.0,
      "cocokp_upsample":1,
      "cocokp_min_kp_anns":1,
      "cocokp_eval_test2017":false,
      "cocokp_eval_testdev2017":false,
      "coco_eval_annotation_filter":true,
      "coco_eval_long_edge":641,
      "coco_eval_extended_scale":false,
      "coco_eval_orientation_invariant":0.0,
      "cifar10_root_dir":"data-cifar10/",
      "cifar10_download":false,
      "save_all":null,
      "show":false,
      "image_width":null,
      "image_height":null,
      "image_dpi_factor":1.0,
      "image_min_dpi":50.0,
      "show_file_extension":"png",
      "show_box":false,
      "white_overlay":false,
      "show_joint_scales":false,
      "show_joint_confidences":false,
      "show_decoding_order":false,
      "show_frontier_order":false,
      "show_only_decoded_connections":false,
      "video_fps":10,
      "video_dpi":100,
      "debug_cifhr":false,
      "debug_cif_c":false,
      "debug_cif_v":false,
      "debug_cifdet_c":false,
      "debug_cifdet_v":false,
      "debug_caf_c":false,
      "debug_caf_v":false,
      "debug_indices":[
         
      ],
      "debug_images":false,
      "output":"outputs/resnet50-201225-112347-cocokp.pkl",
      "stride_apply":1,
      "epochs":75,
      "val_interval":1,
      "train_batches":null,
      "val_batches":null,
      "fix_batch_norm":false,
      "ema":0.01,
      "clip_grad_norm":0.0,
      "log_interval":11,
      "disable_cuda":false,
      "profile":null,
      "device":"cuda",
      "pin_memory":true
   },
   "version":"0+unknown",
   "plugin_versions":{
      
   },
   "hostname":"03c6f42527ad"
}
```
- Evaluation results
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.271
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.497
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.247
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.248
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.319
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.311
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.514
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.302
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.273
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.366
```
- Expected results
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.575
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.798
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.609
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.563
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.605
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.614
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.814
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.649
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.596
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.645
```

So there is a huge delta between evaluation results, I will start next experiment by increasing number of `epochs=150` this will probably produce similar but not exact results.

In general I am interested in retraining model for specific release starting from `v11+`

**Comment by svenkreiss** [comment]
Hello @vladimirmujagic ,

You are not running the correct command. The first line of the log (https://github.com/vita-epfl/openpifpaf-torchhub/releases/download/v0.12a7/resnet50-201123-175351-cocokp-o10s.pkl.log) contains this:

```
"argv": ["/home/kreiss/tech/pifpaf_run/venv3/lib/python3.7/site-packages/openpifpaf/train.py", "--lr=0.0001", "--momentum=0.95", "--b-scale=3.0", "--epochs=250", "--lr-decay", "230", "240", "--lr-decay-epochs=10", "--lr-warm-up-start-epoch=150", "--batch-size=32", "--weight-decay=1e-5", "--dataset=cocokp", "--cocokp-upsample=2", "--cocokp-orientation-invariant=0.1", "--cocokp-extended-scale", "--checkpoint", "outputs/resnet50-201119-121553-cocokp-c2bbdd50.pkl"]
```

These are all the command line parameters that I used to train this model. The training process is similar to shufflenets where the first 150 epochs are trained with the same parameters but without extended-scale and without orientation-invariant. 

I don't know where your "expected results" comes from, but a Resnet50 model in OpenPifPaf is way better than those numbers. 

One of the first things in 2021 will be the release of v0.12. Hopefully, some confusion between old way and dev branch gets resolved.

**Comment by vladimirmujagic** [comment]
Thank you for detailed explanation, I didn't see attached logs.

"Expected results" are produced by cloning openpifpaf repo and checking out `dev` branch commit `6385c33b2c32e305d045362e662d732a9dd3c78a` then running 
`python3 openpifpaf.eval --checkpoint resnet50 --dataset cocokp` this command downloads pretrained model associated with that `branch/commit` and produces posted results.

Results might be related to python and system environment?

**Comment by svenkreiss** [comment]
Please use the command from the Guide (https://vita-epfl.github.io/openpifpaf/dev/train.html <https://vita-epfl.github.io/openpifpaf/dev/train.html>):

```
CUDA_VISIBLE_DEVICES=0 python3 -m openpifpaf.eval \
  --checkpoint resnet50 \
  --loader-workers=2 \
  --decoder=cifcaf:0 \
  --dataset=cocokp --force-complete-pose --seed-threshold=0.2
```

COCO metrics do not penalize extra keypoints for poses, so you should always use --force-complete-pose for this metric. The metric also does not penalize extra poses as long as their instance score is lower than the other poses. Therefore, use a lower seed threshold than normal.
